### ISSUE TITLE
Refactor to proper MVC design

### ISSUE THREAD

**Issue by tonihele** [tonihele/OpenKeeper#263]
Refactor to proper MVC design. Or entity design, it is quite the same also. We are now marrying our M with C. Separate them. And indeed the Paul's (pspeed) ES system might be of use. I think we already have it imported, it might provide some automatic stuff for the network play, And there are some tools that might be better than our hand crafted ones in the package. I haven't looked the examples much but no over-designing! Not every little thing needs to be an entity IMHO.

This will help clean out the code and at the same time it is clearer to implement the creature/trap/object actions. They are now bound to the animations, instead they should be just i.e. damage per second. I don't think the animations are synced really, just to sounds. I don't know how the fighting works in that aspect.

**Comment by tonihele** [comment]
When I started I just wanted to display some graphics, to see if it was possible. Well, it is. Now with the added logic, the design doesn't really support many needed features and is no longer expandable without this redesign.

Goal is to decouple the logic from the graphics. Enable network play and save + load game state. Much like JSettlers have done (I haven't actually looked in to their code, just skimmed the packages they have). Here is also what mr. PSpeed thinks as one possible proper architecture: http://simsilica.github.io/SimEthereal/.

This is quite a huge undertaking but it can't wait much longer.

**Comment by belgianguy** [comment]
pspeed has some other examples as well: https://github.com/Simsilica/Examples

**Comment by tonihele** [comment]
Yeah, from what I can gather Ethreal & Zay-ES are smart to marry together

**Comment by belgianguy** [comment]
True, might be interesting to review pspeed's AsteroidPanic (https://github.com/jMonkeyEngine-Contributions/zay-es/tree/master/examples/AsteroidPanic) as well then. (Found it mentioned here: https://hub.jmonkeyengine.org/t/to-entity-or-not-to-entity/37714, which is another dicussion related to mixed code and graphics)

**Comment by tonihele** [comment]
https://github.com/jMonkeyEngine-Contributions/zay-es/tree/master/examples/MonkeyTrap seems better example out of the two (vs. AsteroidPanic). Just looking at the code. Structurewise...

**Comment by tonihele** [comment]
I was thinking... I'm not treating map as any entity etc. And rooms were really my concern. They too would not fit to any entity. If synced over the network they would need to go with the same tileUpdate stuff to be synced in relation to map state. Feels kinda clumsy. Not maybe that hard.

But I think we can get away without letting the clients know anything about the rooms. They just draw the map. All the fixed buildings in the map are non removable (meaning those with Direction) and readable from the level file. Dungeon heart is basically the only thing that needs to know a state. Basically is it new and is it destroyed. Both with can be specially handled by onPlayerDieded and plug anim as constructing param when loading a new map (since at the server we don't care how it is represented). Super simple. The clients would still hold the instance data for the rooms, so the health etc. they can figure out from the tiles if needed to and can zoom the camera and calculate the types.

This poses problems to acquiring tooltip data perhaps. Like how much of the room capacity is in use etc. Client does have the objects and their state and all that, but doesn't feel very nice. Any suggestions? Maybe also we could now go with a simple model that indeed no rooms to client and change it later once we got the game going....

**Comment by tonihele** [comment]
Actually it is super easy to create a super thin room model. And if I use `long` ID on the room model and add the ID to the tiles it occupies, instead of just using the coordinates to recognize rooms... There is no syncing problem. Alles gut....

**Comment by ufdada** [comment]
Sounds good, you are making some great progress here as far as i can tell from your last commits.

Do you plan to refactor parts of the main menu as well or is it safe to make changes there? I might have some time to finish the campaign selection soon, so i am quite curious.

> Alles gut....

I didn't know you speak german ;)

**Comment by tonihele** [comment]
Thanks, well, very slowly, I'll try my best not to side track. Almost implemented map transferring, UPNP etc. :) But hopefully things will speed up after that and everybody would feel that it is easier to understand & modify the code.

Go ahead. I've done anyway many things on the side now, the old architecture is still there almost intact. So small syncs even there should be ok. Main menu at some point needs to load the map using new loader but that is another thing which I was a bit afraid of. Need to put my thinking hat on, but first the game.

Studied Deutsch for 3-5 years, can't remember anymore, the language or the years :)

**Comment by tonihele** [comment]
Okay, finally :) Quite feature complete with the previous version/design. I tagged the old design just in case there is something done more elegantly or done in the first place and not anymore.

Many features were implemented and/or improved on. Basically this design is also now more stable and less prone to problems caused by modifying the scene graph from outside threads. Logic & view have been separated and game logic runs on a single thread mostly. New stuff have been implemented with ES design and logic is run in sequence by Systems that react to the entities and components. This has also enabled multiplayer,

I made a lot of stuff via interfaces rather than concrete classes. Lot of TODOs and FIXMEs still. Since a lot of work was also keeping much of the old to keep track of the conversion process. The TaskManager is still a bit iffy and produces some errors but mostly working.

What is missing:

- Research, I didn't re-implement it

  - Mainly since the whole player data and room stuff doesn't have a good design at the moment to be honest

- Effects

  - I didn't figure out a definitive good way to implement this, some could be triggered as changes in some components / tiles. Some may be better off using a event bus of sorts
